File: components/visualization/src/main/java/org/datacleaner/visualization/ScatterGroup.java
/**
 * DataCleaner (community edition)
 * Copyright (C) 2014 Neopost - Customer Information Management
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU
 * Lesser General Public License, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this distribution; if not, write to:
 * Free Software Foundation, Inc.
 * 51 Franklin Street, Fifth Floor
 * Boston, MA  02110-1301  USA
 */
package org.datacleaner.visualization;

import java.awt.Point;
import java.io.Serializable;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

import org.datacleaner.api.InputRow;
import org.datacleaner.storage.RowAnnotation;
import org.datacleaner.storage.RowAnnotationFactory;

/**
 * Represents a group of scattered points
 */
public class ScatterGroup implements Serializable {

    private static final long serialVersionUID = 1L;

    private final String _name;

    private final RowAnnotationFactory _rowAnnotationFactory;
    private final Map<Point, RowAnnotation> _annotations = new LinkedHashMap<>();

    public ScatterGroup(String name, RowAnnotationFactory rowAnnotationFactory) {
        _name = name;
        _rowAnnotationFactory = rowAnnotationFactory;
    }

    public String getName() {
        return _name;
    }

    public RowAnnotationFactory getRowAnnotationFactory() {
        return _rowAnnotationFactory;
    }

    public RowAnnotation annotations(Point point) {
        if (_annotations.containsKey(point)) {
            return _annotations.get(point);
        } else {
            final RowAnnotation annotation = _rowAnnotationFactory.createAnnotation();
            _annotations.put(point, annotation);
            return annotation;
        }
    }

    public void register(Number x, Number y, InputRow row, int distinctCount) {
        final Point point = new Point(x.intValue(), y.intValue());
        final RowAnnotation annotation = annotations(point);
        _rowAnnotationFactory.annotate(row, distinctCount, annotation);
    }

    public RowAnnotation getRowAnnotation(Number x, Number y) {
        final Point searchedPoint = new Point(x.intValue(), y.intValue());
        return _annotations.get(searchedPoint);
    }

    public Set<Point> getCoordinates() {
        return _annotations.keySet();
    }

    public Map<Point, RowAnnotation> getAnnotations() {
        return _annotations;
    }

    @Override
    public String toString() {
        return "Name=" + _name + "annotations" + _annotations.keySet();
    }
}

File: components/visualization/src/main/scala/org/datacleaner/visualization/ScatterGroup.scala
package org.datacleaner.visualization

import org.apache.commons.lang3.tuple.{ImmutablePair, Pair}
import org.datacleaner.api.InputRow
import org.datacleaner.storage.{RowAnnotation, RowAnnotationFactory}

import scala.collection.JavaConverters._
import scala.collection.mutable.{LinkedHashMap, Map}

/**
  * Represents a group of scattered points
  */
@SerialVersionUID(1746285060428286068L)
class ScatterGroup(name: String, rowAnnotationFactory: RowAnnotationFactory) extends IScatterGroup {

  def getName: String = name

  val annotations: Map[(Number, Number), RowAnnotation] = LinkedHashMap[(Number, Number), RowAnnotation]().withDefault(
    key => {
      val annotation = rowAnnotationFactory.createAnnotation();
      annotations.put(key, annotation);
      annotation;
    });

  def register(x: Number, y: Number, row: InputRow, distinctCount: Int) = {
    val annotation = annotations((x, y));
    rowAnnotationFactory.annotate(row, distinctCount, annotation);
  }


  def getRowAnnotationFactory = rowAnnotationFactory;

  def getRowAnnotation(x: Number, y: Number) = annotations.getOrElse((x,y), null)

  @transient
  var javaAnnotations : java.util.Map[Pair[Number, Number], RowAnnotation] = null

  def getRowAnnotations : java.util.Map[Pair[Number, Number], RowAnnotation] = {
    this.synchronized {
      if (javaAnnotations == null) {
        javaAnnotations = new java.util.HashMap[Pair[Number, Number], RowAnnotation]()

        annotations.foreach(entry => {
          javaAnnotations.put(new ImmutablePair[Number, Number](entry._1._1, entry._1._2), entry._2)
        })
      }
    }
    javaAnnotations
  }

  def getCoordinates: java.lang.Iterable[Pair[Number, Number]] = {
    val annotations = getRowAnnotations
    val javaCoordinates = new java.util.HashSet[Pair[Number, Number]]();

    annotations.asScala.foreach(entry => {
          javaCoordinates.add(new ImmutablePair[Number, Number](entry._1.getLeft, entry._1.getRight))
        }
    )
    javaCoordinates
  }
}
